Makefile.common

```diff
@@ -8,21 +8,62 @@ define newline
 
 endef
 
-HACL_HOME 	?= $(shell pwd)
+ifndef HACL_HOME
+HACL_HOME 	:= $(shell pwd)
+endif
 
 # Put your local configuration (e.g. HACL_HOME, KRML_HOME, etc.) in
 # Makefile.config
 -include $(HACL_HOME)/Makefile.config
 
-KRML_HOME 	?= $(HACL_HOME)/../karamel
-FSTAR_HOME	?= $(HACL_HOME)/../FStar
-VALE_HOME	?= $(HACL_HOME)/../vale
+ifndef KRML_HOME
+# assuming an Everest source tree
+KRML_HOME 	:= $(HACL_HOME)/../karamel
+endif
+
+ifndef FSTAR_HOME
+  FSTAR_EXE=$(shell which fstar.exe)
+  ifeq ($(FSTAR_EXE),)
+    # F* not found in PATH, assuming an Everest source tree
+    FSTAR_HOME := $(HACL_HOME)/../FStar
+  else
+    # F* found in PATH, assuming directory name ends with /bin
+    FSTAR_HOME := $(dir $(FSTAR_EXE))/..
+  endif
+endif
+
+ifndef FSTAR_ULIB
+FSTAR_ULIB := $(shell if test -d "$(FSTAR_HOME)/ulib" ; then echo "$(FSTAR_HOME)/ulib" ; else echo "$(FSTAR_HOME)/lib/fstar" ; fi)
+endif
+
+ifndef VALE_HOME
+# assuming an Everest source tree
+VALE_HOME	:= $(HACL_HOME)/../vale
+endif
+
+# Make all paths absolute, so that `make` only sees a single view of
+# each file name, especially when generating dependency trees. On
+# Cygwin, we additionally need to use Windows paths instead of Cygwin
+# paths.
+HACL_HOME  := $(realpath $(HACL_HOME))
+KRML_HOME  := $(realpath $(KRML_HOME))
+FSTAR_HOME := $(realpath $(FSTAR_HOME))
+FSTAR_ULIB := $(realpath $(FSTAR_ULIB))
+VALE_HOME  := $(realpath $(VALE_HOME))
+ifeq ($(OS),Windows_NT)
+  HACL_HOME  := $(shell cygpath -m $(HACL_HOME))
+  KRML_HOME  := $(shell cygpath -m $(KRML_HOME))
+  FSTAR_HOME := $(shell cygpath -m $(FSTAR_HOME))
+  FSTAR_ULIB := $(shell cygpath -m $(FSTAR_ULIB))
+  VALE_HOME  := $(shell cygpath -m $(VALE_HOME))
+endif
```

Comment:
I'm generally not a big fan of GNU make functions, but this might be one of the few cases where they might actually make the code more readable, see https://www.gnu.org/software/make/manual/html_node/Call-Function.html -- it's a small nit

Reply:
> I'm generally not a big fan of GNU make functions, but
That's exactly what I was expecting! Let me fix that now.

Reply:
Done in 03543fdc3e5a80a269eb079c7afb455690abc77f